test/e2e/ratelimit/ratelimit_universal.go

```diff
@@ -67,43 +73,62 @@ var _ = E2EBeforeSuite(func() {
 })
 
 func RateLimitOnUniversal() {
-	verifyRateLimit := func(client string, total int) func() int {
-		return func() int {
+	// In the best case, we start sending requests and can send all successfully
+	// and consecutively before the bucket is empty.
+	// Otherwise the bucket empties before we've sent all which will lead to the
+	// number of successes being more than the limit.
+	// So we send requests until the rate limit is first hit, send failing
+	// requests until the first one succeeds, then we send as
+	// many requests as we can until we are rate limited again.
+	// We compare that number of successes to the expected number.
+	// The assumption here is that we can send enough requests to hit the limit
+	// within the interval.
+	verifyRateLimit := func(client string, svc string, expected int) {
+		verifyError := fmt.Sprintf("couldn't verify rate limit for %s to %s to equal %d", client, svc, expected)
+
+		// This is wrapped in Eventually. Shortly after new
+		// RateLimits are applied, there appears to be some nondeterminism around how
+		// many tokens are put in the bucket at one time.
+		verify := func() error {
+			begin := time.Now()
 			succeeded := 0
-			for i := 0; i < total; i++ {
-				_, _, err := cluster.Exec("", "", client, "curl", "-v", "--fail", "test-server.mesh")
-				if err == nil {
+			var haveFailedRequest bool
+			for {
+				_, _, err := cluster.Exec("", "", client, "curl", "-v", "--fail", fmt.Sprintf("%s.mesh", svc))
+				if err != nil {
+					haveFailedRequest = true
+					if succeeded > 0 {
+						// we've failed, then succeeded and are now failing again
+						if succeeded == expected {
+							return nil
+						}
+						return fmt.Errorf("%s: sent %d successful requests", verifyError, succeeded)
+					}
+				} else if haveFailedRequest {
+					// We've failed at least once and have n >= 0 consecutive,
+					// successful requests before this one
 					succeeded++
 				}
+				if time.Now().After(begin.Add(2 * limitInterval)) {
+					return fmt.Errorf("%s: couldn't send a terminated string of successful requests", verifyError)
+				}
 			}
-
-			return succeeded
 		}
-	}
 
-	verifyRateLimitExternal := func(client string, total int) func() int {
-		return func() int {
-			succeeded := 0
-			for i := 0; i < total; i++ {
-				_, _, err := cluster.Exec("", "", client, "curl", "-v", "--fail", "external-service.mesh")
-				if err == nil {
-					succeeded++
-				}
-			}
+		const verifyTimeout = 30 * time.Second
```

Comment:
Maybe express this as `3 * limitInterval`

Reply:
So this timeout interval interface kind of bothers me. The interval is an overall limit on the `Eventually` call IIUC, we don't know how long each try takes, plus the poll interval in between. It's not really related to `limitInterval`. What I actually want is a limit on the number of attempts.